### PR TITLE
refactor: add interfaces

### PR DESCRIPTION
**Pull request by horsefacts**
## Motivation

Minor, but this should make it easier to build on top of the various registries. Allows for it to be an easy reference for the interfaces and can be pulled into other projects to use, rather than needing to redefine it per project.

## Change Summary

Simply adding interfaces. Main change is around where structs are located. 

There are two options, not sure which one you prefer:
1. Structs within the interface
2. Structs in a `*Structs.sol` file and imported into the interface and implementation.

## Merge Checklist

- [x] The PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] The PR has been tagged with change type label(s) (i.e. documentation, feature, bugfix, or chore)
- [x] The PR's changes adhere to all the requirements in the [contribution guidelines](https://github.com/farcasterxyz/contracts/blob/main/CONTRIBUTING.md#3-proposing-changes)
- [x] All [commits have been signed](https://github.com/farcasterxyz/contracts/blob/main/CONTRIBUTING.md#22-signing-commits)

---

## PR-Codex overview
This PR focuses on adding interface contracts and updating existing contracts to implement these interfaces. 

### Detailed summary
- Added `IStorageRegistry` interface to `StorageRegistry.sol`
- Added `IIdRegistry` interface to `IdRegistry.sol`
- Added `IKeyRegistry` interface to `KeyRegistry.sol`
- Added `IBundler` interface to `Bundler.sol`
- Updated imports and function signatures to use the new interfaces

> The following files were skipped due to too many changes: `src/KeyRegistry.sol`, `test/KeyRegistry/KeyRegistry.t.sol`, `src/interfaces/IIdRegistry.sol`, `src/interfaces/IKeyRegistry.sol`, `src/interfaces/IStorageRegistry.sol`, `test/Bundler/Bundler.t.sol`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by sds — COMMENTED** [review]
Looks good to me, but will let @varunsrin make the final call on whether this gets merged.

File: src/interfaces/IBundler.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    /// @notice Data needed to trusted register a user with the fid and storage contracts.
+    struct UserData {
+        address to;
+        address recovery;
+        SignerData[] signers;
+        uint256 units;
+    }
+
+    /// @notice Data needed to trusted register a signer with the key registry
```
Comment (varunsrin) [inline src/interfaces/IBundler.sol:17]: nit: KeyRegistry

```suggestion
    /// @notice Data needed to trusted register a signer with the KeyRegistry
```

File: src/interfaces/IBundler.sol [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+     * @notice Contract version specified in the Farcaster protocol version scheme.
+     */
+    function VERSION() external view returns (string memory);
+
+    /*//////////////////////////////////////////////////////////////
+                                 REGISTRATION
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Register an fid, multiple signers, and rent storage to an address in a single transaction.
```
Comment (varunsrin) [inline src/interfaces/IBundler.sol:57]: nit: can be called by anyone during the Registrable phase.

File: src/interfaces/IIdRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    /**
+     * @notice Maps each fid to an address that can initiate a recovery.
+     */
+    function recoveryOf(uint256 fid) external view returns (address recovery);
+
+    /*//////////////////////////////////////////////////////////////
+                             REGISTRATION LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
```
Comment (varunsrin) [inline src/interfaces/IIdRegistry.sol:52]: is the desired outcome to duplicate comments across the interface and implementation?

this seems like it has the potential to get out of date over time

File: src/interfaces/IIdRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+     * @param sig      EIP-712 Transfer signature signed by the to address.
+     */
+    function recover(address from, address to, uint256 deadline, bytes calldata sig) external;
+
+    /**
+     * @notice Transfer the fid owned by the from address to another address that does not
+     *         have an fid. Caller must provide two signed Transfer messages: one signed by
+     *         the recovery address and one signed by the to address.
+     *
+     * @param from             The owner address of the fid to transfer.
```
Comment (varunsrin) [inline src/interfaces/IIdRegistry.sol:154]: nit: bring this in line with comment above

```suggestion
     * @param from             The address that currently owns the fid.
```

File: src/interfaces/IIdRegistry.sol (hunk 3 above)
Comment (horsefacts) [reply src/interfaces/IIdRegistry.sol:52]: Yeah, we should use the `@inheritdoc` natspec tag.

**Pull request merged**